mp4parse/src/lib.rs

```diff
@@ -2286,6 +2465,184 @@ fn read_schm<T: Read>(src: &mut BMFFBox<T>) -> Result<SchemeTypeBox> {
+                };
+            },
+            BoxType::TempoEntry => meta.beats_per_minute = read_u8_data(&mut b)?
+                .and_then(|tmpo| tmpo.get(1).copied()),
+            BoxType::CompilationEntry => meta.compilation = read_bool_data(&mut b)?,
+            BoxType::AdvisoryEntry => meta.advisory = read_u8_data(&mut b)?
+                .and_then(|rtng| {
+                    Some(match rtng.get(0)? {
+                        2 => AdvisoryRating::Clean,
+                        0 => AdvisoryRating::Inoffensive,
+                        r => AdvisoryRating::Explicit(*r),
+                    })
+                }),
+            BoxType::MediaTypeEntry => meta.media_type = read_u8_data(&mut b)?
+                .and_then(|stik| {
+                    Some(match stik.get(0)? {
+                        0 => MediaType::Movie,
+                        1 => MediaType::Normal,
+                        2 => MediaType::AudioBook,
+                        5 => MediaType::WhackedBookmark,
+                        6 => MediaType::MusicVideo,
+                        9 => MediaType::ShortFilm,
+                        10 => MediaType::TVShow,
+                        11 => MediaType::Booklet,
+                        s => MediaType::Unknown(*s)
+                    })
+                }),
+            BoxType::PodcastEntry => meta.podcast = read_bool_data(&mut b)?,
+            BoxType::TVSeasonNumberEntry => meta.tv_season = read_u8_data(&mut b)?
+                .and_then(|tvsn| tvsn.get(3).copied()),
+            BoxType::TVEpisodeNumberEntry => meta.tv_episode_number = read_u8_data(&mut b)?
+                .and_then(|tves| tves.get(3).copied()),
+            BoxType::GaplessPlaybackEntry => meta.gapless_playback = read_bool_data(&mut b)?,
+            BoxType::CoverArtEntry => meta.cover_art = read_multiple_u8_data(&mut b).ok(),
+            _ => skip_box_content(&mut b)?,
+
+        };
+        check_parser_state!(b.content);
+    }
+    Ok(())
+}
+
+fn read_bool_data<T: Read>(src: &mut BMFFBox<T>) -> Result<Option<bool>> {
+    Ok(read_u8_data(src)?.and_then(|d| Some(d.get(0)? == &1)))
+}
+
+fn read_string_data<T: Read>(src: &mut BMFFBox<T>) -> Result<Option<String>> {
+    read_u8_data(src)?
+        .map_or(Ok(None),
+                |d| String::from_utf8(d)
+                    .map_err(From::from)
+                    .map(Some)
+        )
+}
+
+
+fn read_u8_data<T: Read>(src: &mut BMFFBox<T>) -> Result<Option<Vec<u8>>> {
+    let mut iter = src.box_iter();
+    let mut data = None;
+    while let Some(mut b) = iter.next_box()? {
+        match b.head.name {
+            BoxType::MetadataItemDataEntry => {
+                data = Some(read_data(&mut b)?);
+            }
+            _ => skip_box_content(&mut b)?,
+        };
+        check_parser_state!(b.content);
+    }
+    Ok(data)
+}
+
+
+fn read_multiple_u8_data<T: Read>(src: &mut BMFFBox<T>) -> Result<Vec<Vec<u8>>> {
```

Comment:
It'd be nice if we could make most of read_multiple_u8_data and read_u8_data common.